src/nvim/getchar.c

```diff
@@ -1577,6 +1577,73 @@ vungetc ( /* unget one character (can only be done once!) */
   old_mouse_col = mouse_col;
 }
 
+static void do_insert_message(const int mode_deleted, const int c,
+                              const int advance)
+{
+  /*
+   * The "INSERT" message is taken care of here:
+   *	 if we return an ESC to exit insert mode, the message is deleted
+   *	 if we don't return an ESC but deleted the message before, redisplay it
+   */
+  if (advance && p_smd && msg_silent == 0 && (State & INSERT)) {
+    if (c == ESC && !mode_deleted && !no_mapping && mode_displayed) {
+      if (typebuf.tb_len && !KeyTyped)
+        redraw_cmdline = TRUE;              /* delete mode later */
+      else
+        unshowmode(FALSE);
+    } else if (c != ESC && mode_deleted) {
+      if (typebuf.tb_len && !KeyTyped)
+        redraw_cmdline = TRUE;              /* show mode later */
+      else
+        showmode();
+    }
+  }
+}
+
+static long calc_waittime(const int keylen)
+{
+  if (typebuf.tb_len == 0) {
+    return -1L;
+  } else if (!p_timeout) {
+    return -1L;
+  } else if (!p_ttimeout && keylen == KEYLEN_PART_KEY) {
+    return -1L;
+  } else if (keylen == KEYLEN_PART_KEY && p_ttm >= 0) {
+    return p_ttm;
+  }
+
+  return p_tm;
+}
+
+static int handle_int(const int advance)
```

Comment:
While I appreciate that `int` is often understood as interrupt and brevity is important, I think it would make sense to name this one `handle_interrupt`.